### ISSUE TITLE
Checking session validity using the Provider (Session Management)

### ISSUE THREAD

**Issue by jorgenkg** [panva/node-oidc-provider#171]
I'm trying to implement an SSO wrapped around an external OAuth solution, using `node-oidc-provider`. 

Ideally I would love to use the Session Management's user cookie to check whether the current login session is valid (to spare the end user from having to type their credentials again). 

I've tried to dig into the internals of the koa router, but the code is unluckily to too efficient/performant for me to understand the flow of calls made when the user's browser visits `/session/check`.

TL;DR 

Do you think I can manage to create a hack that lets me verify the user's session on the *oidc-provider server* based on the user's cookie? If so, may I please be nudged in the direction of the relevant source code files?

**Comment by panva** [comment]
Isn't the supported [Session Management](https://openid.net/specs/openid-connect-session-1_0.html) not exactly what you're looking for?

**Comment by jorgenkg** [comment]
I believe Session Management allows a third party client to validate whether a session is active using a redirect/iframe?

While I'm hoping for a way to check whether the user's browser is currently in a active session based on the cookie. The Session Management might indeed be used to do the session validation, but I'm not sure how to call the internals of the Session Management from within the same node as is running `node-oidc-provider`. 

Please refer to the attached image. I would really love to be able to do the quick check "is the user working from an active session right now?".

![sequence diagram](https://user-images.githubusercontent.com/1458231/32279633-9d1ba3e6-bf19-11e7-886d-9d8188f4ff38.png)

**Comment by panva** [comment]
This is what Session Management is for and there's no server call or internals involved, only RP and OP frames.

**Comment by panva** [comment]
Why are you mixing in Session Management then? The description you are providing is massively confusing and lacks detail. Your diagram says Client, which i assume now is not an RP either.

Anyway. Few issues i see with your fn. 
- Unless `session.expiresAt` is something you provide on your own then there's no such property present and it sure is not a Date object.
- can't really say what doesnt work from this end since there is 0 logs included as to where does the function end up. Does the created context have cookies? etc.

**Comment by jorgenkg** [comment]
I didn't manage to sufficiently describe what I desired to do. Nevertheless, if someone else comes across this GitHub issue:

```
function findUserFromSession( provider, req, res ){
  return provider
    .Session
    .get( provider.app.createContext(req, res) )
    .then((session)=>{
      if( session.account !== undefined && new Date() < session.expiresAt){
        return provider.Account.findById( null, session.account, null);  
      }
      
      return null;
    });
}
```

If you believe this is logic is flawed, please point out my mistakes

**Comment by panva** [comment]
Why are you mixing in Session Management then? The description you are providing is massively confusing and lacks detail. Your diagram says Client, which i assume now is not an RP either.

Anyway. Few issues i see with your fn. 
- Unless `session.expiresAt` is something you provide on your own then there's no such property present and it sure is not a Date object.
- can't really say what doesnt work from this end since there is 0 logs included as to where does the function end up. Does the created context have cookies? etc.

**Comment by jorgenkg** [comment]
The utility function above works. It retrieve and validate sessions as expected. The utility itself is merely a re-purposing of internal functions from the excellent *iodc-provider* library. I just wanted to share the function in case it was helpful to others.

1. Why I mentioned Session Management: 
I assumed the library implemented functions to validate whether the cookie (ctx.cookies) of a RP contained a valid session. I was hoping to avoid using postMessages and found a method in the Session model that appeared to do the job.
2. Regarding `session.expiresAt`:
It seems like the library provides an expiresAt on the session? When logging the session object retrieved by `Session.get` it prints out:
```
Session {
  expiresAt: 1541429974529
  ...
}
```
Furthermore, I guessed that the value *expiresAt* might be an unix epoch. When comparing Date objects, they are compared by their `valueOf() -> unix epoch `.

**Comment by panva** [comment]
You’re offering this to RPs? With no authz or proof it authorized before? 

I don’t understand the need to avoid using to be standardized and supported way which takes privacy and security into consideration and requires no server communication on top of it.
